general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+   {
+      PointS best_candidate;
+      best_candidate.sp=pt;
+      //initialize the best candidate
+      best_candidate.pos =0;
+      best_candidate.dist=Dist(data[0].pt, best_candidate.sp);
+      best_candidate.level=0;
+      PSearch(data.begin(), data.end(), 0, best_candidate);
+
+      ind=data[best_candidate.pos].ind;
+      dist=best_candidate.dist;
+   }
+
+   /// Returns the closest point and the distance to the input point pt.
+   void FindClosestPoint(PointND& pt, Tindex& ind, Tfloat& dist,  PointND& clp)
+   {
+      PointS best_candidate;
+      best_candidate.sp=pt;
+      //initialize the best candidate
+      best_candidate.pos =0;
+      best_candidate.dist=Dist(data[0].pt, best_candidate.sp);
+      best_candidate.level=0;
+      PSearch(data.begin(), data.end(), 0, best_candidate);
+
+      ind=data[best_candidate.pos].ind;
+      dist=best_candidate.dist;
+      clp=data[best_candidate.pos].pt;
+   }
+
+
+   /// Brute force search - please, use it only for debuging purposes
+   void rFindClosestPoint(PointND& pt, Tindex& ind, Tfloat& dist)
+   {
+      PointS best_candidate;
+      best_candidate.sp=pt;
+      //initialize the best candidate
+      best_candidate.pos =0;
+      best_candidate.dist=Dist(data[0].pt, best_candidate.sp);
+      Tfloat dd;
+      for (auto iti=data.begin()+1; iti!=data.end(); iti++)
+      {
+         dd=Dist(iti->pt,  best_candidate.sp);
+         if (dd<best_candidate.dist)
+         {
+            best_candidate.pos=iti-data.begin();
+            best_candidate.dist=dd;
+         }
+      }
+
+      ind=data[best_candidate.pos].ind;
+      dist=best_candidate.dist;
+   }
+
+   /// Finds all points within a distance R from point pt. The indices are
+   /// returned in the vestor res and the correponding distances in vector dist.
+   void FindNeighborPoints(PointND& pt,Tfloat R, std::vector<Tindex> & res,
+                           std::vector<Tfloat> & dist)
+   {
+      FindNeighborPoints(pt,R,data.begin(),data.end(),0,res,dist);
+   }
+
+   /// Finds all points within a distance R from point pt. The indices are
+   /// returned in the vestor res and the correponding distances in vector dist.
+   void FindNeighborPoints(PointND& pt,Tfloat R, std::vector<Tindex> & res)
+   {
+      FindNeighborPoints(pt,R,data.begin(),data.end(),0,res);
+   }
+
+   /// Brute force search - please, use it only for debuging purposes
+   void _FindNeighborPoints(PointND& pt,Tfloat R, std::vector<Tindex> & res,
+                            std::vector<Tfloat> & dist)
+   {
+      Tfloat dd;
+      for (auto iti=data.begin(); iti!=data.end(); iti++)
+      {
+         dd=Dist(iti->pt,  pt);
+         if (dd<R)
+         {
+            res.push_back(iti->ind);
+            dist.push_back(dd);
+         }
+      }
+   }
+
+   /// Brute force search - please, use it only for debuging purposes
+   void _FindNeighborPoints(PointND& pt,Tfloat R, std::vector<Tindex> & res)
```

Comment:
Likewise here duplicate code with above `_FindNeighborPoints`.

Reply:
Pointers will make the code more inconvenient.